### PR TITLE
Create tests and classes for pieces

### PR DESCRIPTION
**Pull request by rongg**
Each piece inherits from MyChessPiece and has it's own is_obstructed method. I wrote a few test cases for each piece but I'm sure there are more if you wish to be robust about it. 

I implemented the Bishop's is_obstructed method as an example. It takes an array of Chess pieces, a destination Position, and runs that against the piece(self). Each Chess Piece has a Position, which is another class I created so that we can refer to positions as x and y.

I hope this makes sense! Good night/morning :)

### CODE REVIEW

**Review by ezschemi — CHANGES_REQUESTED** [review]
Wouldn't moving inside_board_boundaries?() and at_destination?() to MyChessPiece make more sense because every chess piece will have these checkers?

File: app/services/bishop.rb [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        puts 'OCCUPIED!'
+        return true
+      end
+    end
+
+    false
+  end
+
+  def inside_board_boundaries?(x, y)
+    x < 8 && x > -1 && y < 8 && x > -1
```
Comment (ezschemi) [inline app/services/bishop.rb]: Can you define a constant for the size of the chess board instead of having magic numbers in the code?

**Review by ezschemi — CHANGES_REQUESTED** [review]
Why is the type-attribute writeable from the outside? It can never change.

Also, the ctor needs to set it, probably passed in as parameter.

**Review by ezschemi — CHANGES_REQUESTED** [review]
Position should have been "Coordinate" or the DB structure document needs to be updated. 

It is missing the definition of where the origin (0, 0) is.

It is also missing constant definitions for places like "not on the board", "in the black/white capture area"

**Review by ezschemi — CHANGES_REQUESTED** [review]
Why is "MyChessPiece" named "MyChessPiece" and not "ChessPiece" ?

**Review by ezschemi — CHANGES_REQUESTED** [review]
MyChessPiece needs an enum for the different types, instead of raw integers.

**Review by ezschemi — CHANGES_REQUESTED** [review]
(My)ChessPiece should have the prototype definition of the is_obstructed-method, and explain what the parameters are.

**Review by ezschemi — CHANGES_REQUESTED** [review]
As there is no documentation for is_obstructed I can only guess: 

Let's say pieces contains two pawns. How do you determine which of the two shall be moved and checked for obstruction?

**Review by ezschemi — CHANGES_REQUESTED** [review]
.should-syntax is deprecated and prints out a warning. Needs to use expect() instead. 
Example for the king:
expect(king.is_obstructed?(pieces, destination)).to eq(true)

**Comment by fleawig** [comment]
Hey guys -- my branch for listing the pieces out -- which we also planned to incorporate the pieces and movements methods we wrote the other day -- has several restructuring changes that are going to conflict with this branch. what fun! I've been trying to finish it all day, just been figuring out how to get the instance variables we wrote to show up in views, since they are now in a module (which don't pass instance variables to views.)

**Comment by fleawig** [comment]
I expect to make that pull request within an hour.

**Comment by fleawig** [comment]
Meanwhile you should really look at the hash constant StartingPositions::STARTING_POSITIONS for how the initial pieces are being handled. (It could be further refactored.)

**Comment by fleawig** [comment]
It may not be necessary to use an enum (maybe not always encouraged in Rails anyway?) if the piece types are built into their names and could be captured that way.

**Comment by rongg** [comment]
I resolved most of the issues brought up by Enrico. Thanks for the good finds :)

For the sake of getting the ball rolling with the is_obstructed methods I'm going to merge this. 
Jesse, I can help in resolving the conflicts created by this merge with your branch.

**Comment by ezschemi** [comment]
When will you resolve the still open issues then?



Enrico





On Fri, Jan 20, 2017, at 06:08, rongg wrote:

> I resolved most of the issues brought up by Enrico. Thanks for the
> good finds :)
> For the sake of getting the ball rolling with the is_obstructed
> methods I'm going to merge this.
>  Jesse, I can help in resolving the conflicts created by this merge
>  with your branch.
> — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub[1], or mute the thread[2].
> 




Links:

  1. https://github.com/thesnowcoders-fh/sprky-chess/pull/9#issuecomment-273978917

**Comment by rongg** [comment]
Based on the commits I believe there are 2 issues not addressed yet:

1)**It is also missing constant definitions for places like "not on the board", "in the black/white capture area"**

I didn't address this because we haven't gotten to capturing pieces yet, but it's probably good to define it. How about captured pieces have two possible positions?

**Captured black pieces:** 
Position (0, -1) # below board

**Captured white pieces:**
Position (0, 8) # above board

To query a certain color's captured pieces look for pieces with these positions
To query if any piece part of the game, call ChessPiece.inside_board_boundaries?


2)**MyChessPiece needs an enum for the different types, instead of raw integers.**

Just thinking of it, I'm not sure we need this field now that we have inheritance. To know what type a piece is we can use reflection but I don't see a scenario where we'd have to do this .

**Comment by ezschemi** [comment]
Ron,



numeric values are not going to fly for the special positions. Either
we need a method for each and/or constant definitions (just like the
board size).


I agree on the second issue, probably not needed anymore.



Enrico







On Fri, Jan 20, 2017, at 20:01, rongg wrote:

> Based on the commits I believe there are 2 issues not addressed yet:
> 1)*It is also missing constant definitions for places like "not on the
>   board", "in the black/white capture area"*
> I didn't address this because we haven't gotten to capturing pieces
> yet, but it's probably good to define it. How about captured pieces
> have two possible positions?
> *Captured black pieces:*

>  Position (0, -1) # below board



> *Captured white pieces:*

>  Position (0, 8) # above board



> To query a certain color's captured pieces look for pieces with these
> positions
>  To query if any piece part of the game, call
>  ChessPiece.inside_board_boundaries?
> 2)*MyChessPiece needs an enum for the different types, instead of raw
>   integers.*
> Just thinking of it, I'm not sure we need this field now that we have
> inheritance. To know what type a piece is we can use reflection but I
> don't see a scenario where we'd have to do this .
> — You are receiving this because you commented. Reply to this email
> directly, view it on GitHub[1], or mute the thread[2].
> 




Links:

  1. https://github.com/thesnowcoders-fh/sprky-chess/pull/9#issuecomment-274151854

**Pull request merged**